### ISSUE TITLE
PreprocWrapper of TuneWrapper

### ISSUE THREAD

**Issue by davidwangz137** [mlr-org/mlr#461]
Hello,

I was trying to wrap a learner that was already wrapped with makeTuneWrapper with makePreprocWrapper. This gave the error:

Error in makeBaseWrapper(id, type = learner$type, next.learner = learner,  : 
  Cannot wrap an optimization wrapper with something else!

I want to have a complex preprocessing step before passing the data to the tuning learner. The reason I need it in a learner format instead of just preprocessing the data beforehand is that I want to use it in makeStackedLearner, where some of the base learners won't have this preprocessing step.

Thanks

**Comment by larskotthoff** [comment]
Have you tried first adding the PreprocWrapper and then the TuneWrapper?

**Comment by davidwangz137** [comment]
I have tried that and it works. However, I would like to do the preprocessing step first to save computation.

**Comment by larskotthoff** [comment]
To do the preprocessing first, you have to add the PreprocWrapper first -- the order in which things are applied is the order in which you wrap things.

**Comment by davidwangz137** [comment]
I added the PreprocWrapper to the learner first. Then I added the tuning wrapper. By the print to console, I believe that the preproc wrapper is executing after the tuning wrapper. Intuition for me is that the last wrapping is the outermost, and is executed first before going down to the deeper layers of wrapping. I'm using mlr version 2.4.

Here is my code:
```{r}
library(mlr)

#Create the self tuning learner
lrn = makeLearner("regr.nnet", trace = FALSE, decay = 1e-02)

#Create the PCA learner

#the only arguments that this gets when tuning is args$num.vectors
train.step1 = function(data, target, args=list(center=TRUE,scale=TRUE)) {
  #Obtain the numeric features for PCA
  print(args)
  cns = colnames(data)
  nums = setdiff(cns[sapply(data, is.numeric)], target)
  #Set the number of eigenvectors to be at maximum the number of numeric parameters
  if(args$num.vectors > length(nums)) args$num.vectors <- length(nums)
  x = as.matrix(data[, nums, drop = FALSE])
  #Obtain the SVD
  y <- prcomp(x,scale=TRUE)
  
  data = data[, setdiff(cns, nums), drop = FALSE]#Keep only the columns which have the non-PCA'ed columns. These are the factor values
  data = cbind(data, as.data.frame(y$x[,1:args$num.vectors]))
  #The projection onto the eigenvectors is no longer needed
  y$x <- NULL
  return(list(data = data, control = y))
}

pred.step1 = function(data, target, args, control) {
  #Obtain the numeric features for PCA
  cns = colnames(data)
  nums = setdiff(cns[sapply(data, is.numeric)], target)
  #Set the number of eigenvectors to be at maximum the number of numeric parameters
  if(args$num.vectors > length(nums)) args$num.vectors <- length(nums)
  
  x = as.matrix(data[, nums, drop = FALSE])
  x <- scale(x,center = control$center, scale = control$scale)
  #Obtain the SVD
  mat.proj <- control$rotation[,1:args$num.vectors]
  data.proj <- x %*% mat.proj#Get the projection of the numeric values onto the PC eigenvectors.
  #Combine the numeric and categorical data together
  data = data[, setdiff(cns, nums), drop = FALSE]
  data = cbind(data, as.data.frame(data.proj))
  return(data)
}

ps = makeParamSet(
        makeIntegerParam("num.vectors",lower=2,upper=12,default=6)
)

step1 = makePreprocWrapper(lrn, train = train.step1, predict = pred.step1, par.vals=list(num.vectors=6),par.set=ps)

#Add in the tuning on top of it

rdesc = makeResampleDesc("CV", iters = 10)

ctrl = makeTuneControlGrid(resolution=2L);

ps.tune = makeParamSet(
        makeIntegerParam("size",lower=5,upper=20),
        makeNumericParam("decay",lower=-4,upper=0,trafo=function(x)10^x)
)

step2 = makeTuneWrapper(step1,rdesc,mse,par.set=ps.tune,ctrl,show.info=T)

```

**Comment by larskotthoff** [comment]
Ah yes, I see what you mean now -- you're right that the preprocessing is run on the inner loop for each specific parameter setting. I misunderstood you before.

The reason you can't wrap an optimization wrapper is that the optimization and the preprocessing could interact, i.e. you could be optimizing for the wrong thing. It doesn't look like your preprocessing would be terribly expensive -- do you find that it slows things much down in practice?

**Comment by davidwangz137** [comment]
The slowdown is not significant. An example significant slowdown is if I wanted to cluster the data and have the cluster number as a feature. This might take a long time depending on the data and algorithm. Since the clustering is independent of the model that I'm training, it should be done before passing the data into the tuning loop. I feel like it should be an option to wrap a tuning wrapper because of this.

I understand how you would want the preprocessing step in the tuning loop, as whatever feature selection algorithm (ex: simulated annealing) will find the best features for that hyperparameter set. Maybe output a warning message telling the user of the potential interaction between preprocessing and optimization causing the model to not perform as they want?

**Comment by larskotthoff** [comment]
Well it's hard to tell in general when these kinds of interactions occur and what effect they have -- a warning message wouldn't really be helpful in this case.

I understand that in your case it would make sense to have the preprocessing happen outside of the tuning -- maybe you can achieve the same by doing the preprocessing on the task instead of in a wrapper?

**Comment by davidwangz137** [comment]
I would do the preprocessing outside of the learner. However, the learner is part of an ensemble created using makeStackedLearner which has other base learners which don't do the preprocessing, so I need the preprocessing to be a part of the base learner itself.

**Comment by larskotthoff** [comment]
I see. @berndbischl will have a better understanding why things can't be wrapped the way you want them to be, so pinging him.

So the short answer is that it's not supported this way round (and as I've said, I wouldn't worry too much about the additional computational cost for what you're doing), but patches welcome if you feel up to it :)

**Comment by berndbischl** [comment]
Hi. I think I have generally understood what you are talking about. 
But let us be more specific. Can you try to describe exactly what you are doing?

Let us assume data D. We have a learner L. L has some params. Therefore you wrap it around with a TuningWrapper. Now each time it gets trained on a training set, the params get autotuned.
So far so good. 

Now you also have a preproc method P. Maybe that also has some params.
 One option we support, apparently you know this, is to add the PreprocWrapper to the learner, that add the preproc before each training step. Then you add the TuningWrapper. This would enable tuning either only the params of the learner (and always doing in principle the same preproc step), or tuning both the learner and the preproc together. 

Another option you have is to apply the preproc step just as a normal R function to you whole data D. For that you dont need any wrappers. It is also very fast, as you only run the preproc.
 Just apply it. But that has many drawbacks

- if you have params in the preproc, you might have no way to select them, as you dont know which are the best ones
- if you manually try out many things on the full data, them crossvalidate again and again (and manually optimize the prepproc settings) your evaluation score might get overly optimistic and bias.

but tell me what you wanna do exactly, best with an example

**Comment by berndbischl** [comment]
Ok, reading your comment above, is it like this:

You have base learners A, B, C, D.
They are supposed all to go into a stack.
A and B shall do preproc P. C and D dont do P before training.
Now you add P with a PreprocWraper to A, and also to B.
When you train the stack, A gets trained, then B.
But now P is called twice, and produces the same result, both for A and B. Which is wasteful in terms of time. Is that the problem?

**Comment by davidwangz137** [comment]
I should have given a concrete example sooner. Hopefully this will clear things up:

I have three learners A,B,C. There are two preprocessing methods P1 and P2. I want to apply P1 to A and P2 to B. I also want to tune A after it gets the data from P1 with tuner T. Same with B. Then I want to stack all of these learners.
      | -> P1 -> T -> A
S --| -> P2 -> T -> B
      | -> C
My problem is that currently you cannot wrap a tuning wrapper, so I can only do this:
      | -> T -> P1 -> A
S --| -> T -> P2 -> B
      | -> C
If I'm not tuning the preprocessor's parameters, then it would be a waste for it to preprocess every tuning iteration. That is the problem.

**Comment by pat-s** [comment]
Cleaning up / closing old issues. I assume that this enhancement is not of interest anymore after all this time. 
Feel free to re-open if s.o. is tackling it again.
